### PR TITLE
feat: quality test for nutriscore on olive oils

### PR DESCRIPTION
**Pull request by stephanegigandet**
Added a test to verify that we estimate the fruits/vegetables content of olive oils without ingredients to 100% for nutriscore computation.

### CODE REVIEW

**Comment by benbenben2** [comment]
What is the difference between this new test (en-olive-oil-no-ingredients) and the already existing test named "olive-oil"?

File: lib/ProductOpener/Tags.pm [hunk 1]
```diff
@@ -1302,6 +1302,36 @@ sub build_tags_taxonomy ($tagtype, $publish) {
...
+						. " is incorrect\n";
+
+					$errors .= "ERROR - " . $msg;
+				}
+			}
+			elsif ($line =~ /^expected_ingredients:en:/) {
+				# the line should contain a single ingredient
+				my $expected_ingredients = $';    # everything after the matched string
+
+				if ($expected_ingredients =~ /,/i) {
```
Comment (benbenben2) [inline lib/ProductOpener/Tags.pm:1324]: @stephanegigandet, is there a function that could be used there to make sure that the ingredient provided is in the ingredients taxonomy?

**Comment by benbenben2** [comment]
Fixes #8353 

1/ Remark: nutriscore grades are a, b, c, d, e letters. Whereas nutriscore scores are some some numbers. Hence, I renamed **expected_nutriscore:en:c** (suggested in the issue) to **expected_nutriscore_grade:en:c**

2/ Remark 2: following ingredients exist: **en:extra-virgin olive oil**, **en:virgin olive oil** and **en:olive oil**. 
Do we really want to expect **en:olive oil** as ingredient of **en:Extra-virgin olive oils** category (not **extra-virgin olive oil**)?

3/ Tried to prevent incorrect tags in Tags.pm (although these errors are ignored)

Capture d’écran du 2023-05-07 00-31-23
______________________________________________________________________________________________________________________________
Capture d’écran du 2023-05-07 00-18-01
______________________________________________________________________________________________________________________________
Capture d’écran du 2023-05-07 00-03-43


4/ Tried to differentiate between cases (missing nutriscore grade, which grade instead of which one)  in the quality error names to allow more "fine-grained" monitoring. @CharlesNepote will that be a burden for data_quality taxonomy? (we can have variants for all categories that will have these 2 tags, different nutriscore grade values) see DataQualityFood.pm.

______________________________________________________________________________________________________________________________
Capture d’écran du 2023-05-06 16-26-35
______________________________________________________________________________________________________________________________
Capture d’écran du 2023-05-06 16-25-49
______________________________________________________________________________________________________________________________

File: lib/ProductOpener/Tags.pm (hunk 1 above)
Comment (stephanegigandet) [reply lib/ProductOpener/Tags.pm:1324]: there's an exist_taxonomy_tag() function, but that would imply that the ingredients taxonomy would need to be loaded and built before the categories taxonomy.

**Comment by stephanegigandet** [comment]
Great! Thanks a lot for the PR! :)

> 2/ Remark 2: following ingredients exist: **en:extra-virgin olive oil**, **en:virgin olive oil** and **en:olive oil**. Do we really want to expect **en:olive oil** as ingredient of **en:Extra-virgin olive oils** category (not **extra-virgin olive oil**)?

I would check that the actual ingredient is either the expected ingredient, or a child of it. There's a function for it:

is_a("ingredients","en:virgin-olive-oil","en:olive-oil")

> 4/ Tried to differentiate between cases (missing nutriscore grade, which grade instead of which one) in the quality error names to allow more "fine-grained" monitoring. @CharlesNepote will that be a burden for data_quality taxonomy? (we can have variants for all categories that will have these 2 tags, different nutriscore grade values) see DataQualityFood.pm.

I would suggest to have only one generic error, that does not specify the category, something like "Nutri-Score does not match the Nutri-Score expected for the category". For monitoring, we can always add a /categories facet. If we put too many things in the tag, then it becomes impossible to translate them.

One thing to note is that we show data errors very prominently in the producers platform, that's why we try to translate them.

File: lib/ProductOpener/Tags.pm (hunk 1 above)
Comment (benbenben2) [reply lib/ProductOpener/Tags.pm:1324]: then, I will leave it as is for now

**Comment by benbenben2** [comment]
@stephanegigandet 

Reduced to 1 error for nutrition grade and 1 error for ingredient

Included is_a:
```
	test 1: olive oil 			   -> no error
	test 2: extra virgin olive oil -> no error
	test 3: (no ingredients)         -> error
	test 4: salt 				   -> error
	test 5: olive oil, olive oil  -> error
	test 6: switch in FR and, huile d'olive ->  no error
```


	nutriscore E -> error
	nutriscore D -> error
	nutriscore C -> no error

**Comment by stephanegigandet** [comment]
Updated the tests results to make them pass.

**Comment by teolemon** [comment]
@stephanegigandet merge conflict

**Comment by benbenben2** [comment]
Thank you @stephanegigandet! I was stuck

**Comment by benbenben2** [comment]
- Merging with main locally, there are no conflicts but "make cover" leads to many errors:

> tests/unit/attributes.t                     (Wstat: 2048 Tests: 11 Failed: 8)
>   Failed tests:  1-2, 4-9
>   Non-zero exit status: 8
> tests/unit/ecoscore.t                       (Wstat: 3840 Tests: 69 Failed: 15)
>   Failed tests:  29-36, 41, 47-48, 53-54, 56, 63
>   Non-zero exit status: 15
> tests/unit/forest_footprint.t               (Wstat: 1024 Tests: 9 Failed: 4)
>   Failed tests:  4, 7-9
>   Non-zero exit status: 4
> tests/unit/ingredients.t                    (Wstat: 4096 Tests: 53 Failed: 16)
>   Failed tests:  1, 3, 15, 17, 24, 32, 38-42, 44, 50-53
>   Non-zero exit status: 16
> tests/unit/ingredients_processing.t         (Wstat: 256 Tests: 61 Failed: 1)
>   Failed test:  61
>   Non-zero exit status: 1
> tests/unit/nutriscore.t                     (Wstat: 1024 Tests: 27 Failed: 4)
>   Failed tests:  9, 15-16, 21
>   Non-zero exit status: 4
> tests/unit/packaging.t                      (Wstat: 5120 Tests: 69 Failed: 20)
>   Failed tests:  12-15, 26, 29, 34, 41-42, 45-46, 50-56
>                 62, 66
>   Non-zero exit status: 20
> tests/unit/recipes.t                        (Wstat: 1280 Tests: 8 Failed: 5)
>   Failed tests:  1-2, 4, 6-7
>   Non-zero exit status: 5

- "make update_tests_results" runs until:

>   Waited too much for backend at /opt/product-opener/lib/ProductOpener/APITest.pm line 128.
>     ProductOpener::APITest::wait_server() called at /opt/product-opener/lib/ProductOpener/APITest.pm line 141
>     ProductOpener::APITest::wait_application_ready() called at integration/search_v1.t line 18
> make: *** [Makefile:291 : update_tests_results] Erreur 22

- Now, "make cover" works fine
- Commit and push
- Now fails with this:

> tests/integration/api_v2_product_read.t          (Wstat: 1280 Tests: 43 Failed: 5)
>   Failed tests:  5, 29, 32, 35, 38
>   Non-zero exit status: 5
> tests/integration/api_v2_product_write.t         (Wstat: 512 Tests: 16 Failed: 2)
>   Failed tests:  5, 11
>   Non-zero exit status: 2
> tests/integration/api_v3_product_read.t          (Wstat: 1280 Tests: 48 Failed: 5)
>   Failed tests:  5, 32, 35, 38, 41
>   Non-zero exit status: 5
> tests/integration/data_quality_knowledge_panel.t (Wstat: 512 Tests: 6 Failed: 2)
>   Failed tests:  2, 5
>   Non-zero exit status: 2
> tests/integration/protected_product.t            (Wstat: 1536 Tests: 34 Failed: 6)
>   Failed tests:  9, 15, 19, 23, 29, 33
>   Non-zero exit status: 6



Need help on this @stephanegigandet @alexgarel @teolemon

**Comment by alexgarel** [comment]
@benbenben2 I pushed 508469898d67aa34270b452192c602840d7a5c03 which:
* fix taxonomy name
* update tests results. 

In your previous version, lots of expected_tests_results had "en:ingredients-single-ingredient-from-category-does-not-match-actual-ingredients" added, without any reason for that (like in cookies for example).

**Comment by benbenben2** [comment]
Thanks a lot @alexgarel :1st_place_medal:

**Pull request merged**